docs/CompilerPerformance.md

```diff
@@ -0,0 +1,1115 @@
+in a problem area, and work on those first.
+
+Sorting only tells you where a cost center is, though, not whether that cost
+center is _reasonable_. That is, it's often helpful to try to differentiate
+cost-centers that "should" be expensive (because they're compiling a lot of
+code) from cost-centers that "shouldn't" be expensive (but are, because they
+have a bug or design flaw). One tool that's useful for differentiating these
+cases is the `utils/scale-test` script.
+
+#### Scale-test
+
+This script works with parametrized templates of code (in the same `gyb` format
+as other generated code in the compiler), repeatedly instantiating the template
+with larger and larger parameter values, and measuring the relationship between
+the input parameter value and the outputs of the compiler counters.
+
+The `utils/scale-test` script works with the same counters inside the compiler
+as the `-Xfrontend -print-stats` option mentioned above, so it should be used
+with an assert compiler. It works in concert with the `utils/gyb` boilerplate
+code-generation tool. The process is straightforward:
+
+  1. Write a small `.gyb` file that contains a code pattern that can be varied
+     by varying a numeric input parameter called `N`. For example, the
+     following file (taken from the testsuite) varies the number of nested
+     dictionaries:
+     ```
+     private let _: [Any?] = [[
+     %for i in range(0, N):
+       "A": [
+         "B" : "C",
+     %end
+     %for i in range(0, N):
+         ]
+     %end
+     ]]
+     ```
+
+  2. Run the file under the `utils/scale-test` script. You will at least need to
+     pass a `--swiftc-binary` and `.gyb` template filename; by default, it will
+     test on values of `N` ranging from `10` to `100` in steps of `10`, and fit
+     scaling curves to _all_ counters that it measures, printing any that scale
+     worse than `O(n^1.2)`. For example, the following will give an initial
+     survey of the script above:
+     ```
+     $ utils/scale-test --swiftc-binary=/.../usr/bin/swiftc test.swift.gyb
+     O(n^0.0) : BasicCalleeAnalysis.computeMethodCallees
+     O(n^0.0) : Clang module importer.NumTotalImportedEntities
+     O(n^0.0) : Constraint solver largest system.LargestNumDisjunctionTerms
+     O(n^0.0) : Constraint solver largest system.LargestNumDisjunctions
+     ...
+     O(n^1.0) : regalloc.NumCopies
+     O(n^1.0) : isel.NumFastIselSuccessIndependent
+     O(n^1.0) : regalloc.NumLoads
+     O(n^1.0) : Constraint solver overall.NumSolutionAttempts
+     O(n^1.9) : Constraint solver overall.TotalNumTypeVariables
+     O(n^2.0) : Constraint solver overall.OverallNumComponentsSplit
+     O(n^2.0) : Constraint solver largest system.LargestNumComponentsSplit
+     ...
+     ```
+
+  3. Focus in on a single problematic counter if you see any, and make a
+     regression scale-test -- with a `// RUN: %scale-test` or `// RUN: not
+     %scale-test` line in it depending on whether it's a positive or negative
```

Comment:
Is there a bug about having a better way for negative tests? E.g. I suspect `not %scale-test` will succeed if the code is just totally invalid and doesn't demonstrate the problem; maybe `%scale-test --expected-nonlinear-polynomial ...` and/or `%scale-test --expected-exponential`, or even `%scale-test --expected-nonlinear-polynomial-range 2-3` and `%scale-test --expected-exponential-range 10-11` meaning the exponent/base (respectively) should be in the range specified, ensuring we can try to even avoid regressions from quadratic to cubic.

Reply:
Yeah, Mark just added a mode for this (#11692), rewrote to reference it.